### ISSUE TITLE
Null pointer exception in Fields.all

### ISSUE THREAD

**Issue by salami** [vestrel00/contacts-android#286]
We have several crash reports in production with similar stack traces that end up calling `Fields.all`. For example:
```
Exception java.lang.NullPointerException:
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.NoteFields.getAll (NoteFields.java)
  at contacts.core.Fields$all$2.invoke (Fields.java:3)
  at contacts.core.Fields$all$2.invoke (Fields.java:3)
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.Fields.getAll (Fields.java)
  at contacts.core.IncludeKt.includeAllFields (Include.kt)
  at contacts.core.PhoneLookupQueryImpl.<init> (PhoneLookupQueryImpl.java:2)
  at contacts.core.PhoneLookupQueryKt.PhoneLookupQuery (PhoneLookupQuery.kt:2)
  at contacts.core.ContactsImpl.phoneLookupQuery (ContactsImpl.java:2)
```
And another
```
Exception java.lang.NullPointerException:
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.DataContactsOptionsFields.getAll (DataContactsOptionsFields.java)
  at contacts.core.DataContactsFields$all$2.invoke (DataContactsFields.java)
  at contacts.core.DataContactsFields$all$2.invoke (DataContactsFields.java)
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.DataContactsFields.getAll (DataContactsFields.java)
  at contacts.core.Fields$all$2.invoke (Fields.java:1)
  at contacts.core.Fields$all$2.invoke (Fields.java:1)
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.Fields.getAll (Fields.java)
  at contacts.core.IncludeKt.includeAllFields (Include.kt)
  at contacts.core.PhoneLookupQueryImpl.<init> (PhoneLookupQueryImpl.java:2)
  at contacts.core.PhoneLookupQueryKt.PhoneLookupQuery (PhoneLookupQuery.kt:2)
  at contacts.core.ContactsImpl.phoneLookupQuery (ContactsImpl.java:2)
```
And another
```
Exception java.lang.NullPointerException:
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.SipAddressFields.getAll (SipAddressFields.java)
  at contacts.core.Fields$all$2.invoke (Fields.java:5)
  at contacts.core.Fields$all$2.invoke (Fields.java:5)
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
  at contacts.core.Fields.getAll (Fields.java)
  at contacts.core.IncludeKt.includeAllFields (Include.kt)
  at contacts.core.PhoneLookupQueryImpl.<init> (PhoneLookupQueryImpl.java:2)
  at contacts.core.PhoneLookupQueryKt.PhoneLookupQuery (PhoneLookupQuery.kt:2)
  at contacts.core.ContactsImpl.phoneLookupQuery (ContactsImpl.java:2)
```

For some reason the lazy initialization is failing. Will report back as I find more.

**Comment by vestrel00** [comment]
Interesting. Weird, 🤣 I'll take a look right now! 

If I can't find the exact cause, I'll just get rid of the lazy init calls. I don't think they are doing that much optimization anyways.

Regardless, this is a critical issues so it will be included in the upcoming [0.3.0 release](https://github.com/vestrel00/contacts-android/discussions/218) 🔥

**Comment by salami** [comment]
Awesome, thanks for the quick response! And yes, it is strange.

**Comment by vestrel00** [comment]
I checked the order of declaration of the fields and I don't think that is it..

So, IMO there is really only one possibility at this point...

Your app is using this library in multi-threaded environments. This should, obviously, be 100% okay to do as the library should work in multi-threaded usages. Otherwise, no one would use this library 😅. 

So, the problem here could be that there is no thread safety in place when initializing fields' `all` values.

Currently, those are initialize without thread safety.

```kotlin
lazy(LazyThreadSafetyMode.NONE) { ... }
```

I thought this would be okay... But perhaps I was wrong...

If this is indeed the issue, then the fix for this is to add some thread safety to field initialization.

```kotlin
lazy(LazyThreadSafetyMode.SYNCHRONIZED) { ... }
// or just lazy { ... }
```

**Comment by vestrel00** [comment]
So, I took a deeper look and I think I know 100% what's going on now. First of all, I'm an idiot! Just had to make that clear first lol.

Okay, this is multi-threading 101 and I completely missed it...

Here's the stack trace you provided,

```kotlin
Exception java.lang.NullPointerException:
  at kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java)
```

Here is the official Kotlin implementation that throws the `NullPointerException`,

![Screenshot 2023-05-24 at 10 10 26 AM](https://github.com/vestrel00/contacts-android/assets/4332753/432c8a62-6775-48e8-b9db-13ff49464d19)

Pay close attention to the following two lines of code,

```kotlin
_value = initializer!!()
initializer = null
```

The `NullPointerException` **must** be getting thrown at the invocation of `initializer!!`.

Here's how it could happen. Let's say that there are two threads X and Y. Since there is no locking/synchronization in place, the code could be executed at runtime in this order...

```kotlin
ThreadX { _value = initializer!!() }
ThreadX { initializer = null }
ThreadY { _value = initializer!!() } // throws NullPointerException
```

**Comment by salami** [comment]
Under certain situations, we are launching more than one coroutine at the same time and inside each we are querying the API.

So I think this is the cause as you suspect. We might be able to get around it by moving the API calls into one single coroutine. I am new to my app's codebase so need to investigate.

The other option I suppose might be having your API allow for configuration to use `LazyThreadSafetyMode.SYNCHRONIZED` instead of `LazyThreadSafetyMode.NONE` as you suggest.

**Comment by salami** [comment]
Oh wow, great find!

**Comment by vestrel00** [comment]
@salami, I will fix this issue now. However, I will not be able to release the [0.3.0](https://github.com/vestrel00/contacts-android/discussions/218) for another week or so because I'm busy at work. 

> The other option I suppose might be having your API allow for configuration to use LazyThreadSafetyMode.SYNCHRONIZED instead of LazyThreadSafetyMode.NONE as you suggest.

No need for consumer input/configuration. This library should be using `LazyThreadSafetyMode.SYNCHRONIZED` regardless. It should not be configurable.

> We might be able to get around it by moving the API calls into one single coroutine. I am new to my app's codebase so need to investigate.

I don't think you need to do this. One quick temporary fix that you can do right now without having to wait for v0.3.0 is to ensure all API fields are initialized before any threads are launched that uses the `Contacts` API or `Fields`.

You could do something as simple as this...

```kotlin
class SalamiApp : Application() {
    override fun onCreate() {
        super.onCreate()
        // Temporary fix for https://github.com/vestrel00/contacts-android/issues/286
        // Remove this when upgrading to contacts-android 0.3.0
        Fields.all
        ...
    }
}
```

**Comment by vestrel00** [comment]
This should be fixed now in the upcoming [0.3.0 release](https://github.com/vestrel00/contacts-android/discussions/218) 🔥 🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥

THANKS A LOT @salami for raising this super important issue 🙏 🙇 ❤️ 

For those that want a fix prior to upgrading to 0.3.0, please add the following line of code prior to using the `Contacts` API in multi-threaded scenarios. It could be as simple as this;

```kotlin
class MyApp : Application() {
    override fun onCreate() {
        super.onCreate()
        // Temporary fix for https://github.com/vestrel00/contacts-android/issues/286
        // Remove this when upgrading to contacts-android 0.3.0
        Fields.all // initialize all API fields in a single thread
        ...
    }
}
```

**Comment by salami** [comment]
Thank you! I am very impressed by your speed at finding a proper solution. Looking forward to v0.3.0.

**Comment by salami** [comment]
Something I was wondering -- does using lazy initialization here make a notable difference in performance? From what I am seeing, the Fields basically are comprised of string literals.

If you're not sure, no worries. No need to run tests or anything. Just curious.

**Comment by vestrel00** [comment]
This is included in [0.3.0](https://github.com/vestrel00/contacts-android/releases/tag/0.3.0)!
